File: chap14_respiration/PIR_HH.m

%  -------------------------------------------------------------------
%
%   Code to simulate post-inhibitory rebound coupled HH neurons.
%
%   For Chapter 14, Section 14.7.2 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
%
%   Written by James Keener and James Sneyd.
%
%  -------------------------------------------------------------------

function PIR_HH
close all
clear all
clc
set(0,                           ...
'defaultaxesfontsize', 20,   ...
'defaultaxeslinewidth', 2.0, ...
'defaultlinelinewidth', 2.0);

global gpir   Vpir gL gsyn Vsyn phi thetasyn ksyn VL

gsyn = 0.3;
thetasyn = -44;
gpir = 0.3;
ksyn = 2;
Vpir = 120;
VL = -60;
Vsyn = -80;
gL = 0.1;
phi = 3;

% integrate the ode
init = [-50;1;-40;0]; %initial data for the odes
tstep = 0.1;
t_end = 300;

%specify the output points
tspan = [-50:tstep:t_end];

[T,S] = ode15s(@(t,x)rhs(t,x),tspan,init);
V1=S(:,1);
V2 = S(:,3);

figure(1)
    plot(T,V1,T,V2,'--','linewidth',2)
    legend('boxoff')
    legend('V_1','V_2')
    xlabel('Time')
    axis([0 300 -80 0])

end % of main

%% the right hand side for ode simulation: 

function out=rhs(t,s)
    global gpir  Vpir gL gsyn Vsyn phi VL
    
    V1 = s(1);
    h1 = s(2);
    V2 = s(3);
    h2 = s(4);
    
    out1 = gates(V1); 
    minf1 = out1(1);
    hinf1 = out1(2);
    tauh1 = out1(3);
    Sinf1 = out1(4);
    
    out2 = gates(V2);
    minf2 = out2(1);
    hinf2 = out2(2);
    tauh2 = out2(3);
    Sinf2 = out2(4);
    
    V1p = -gpir*(minf1^3)*h1*(V1-Vpir) - gL*(V1-VL) - gsyn*Sinf2*(V1-Vsyn);
    h1p = phi*(hinf1-h1)/tauh1 ;
    V2p = -gpir*(minf2^3)*h2*(V2-Vpir) - gL*(V2-VL) - gsyn*Sinf1*(V2-Vsyn);
    h2p = phi*(hinf2-h2)/tauh2 ;
    
    out = [V1p;h1p;V2p;h2p];
end

%%
function out = gates(V)
    global thetasyn ksyn
    minf = 1/(1+exp(-(V+65)/7.8));
    hinf = 1/(1+exp((V+81)/11));
    tauh=hinf*exp((V+162.3)/17.8);
    Sinf = 1/(1+exp(-(V-thetasyn)/ksyn));
    
    out = [minf,hinf,tauh,Sinf];
end

File: chap14_respiration/PIR_HH.py

#   -------------------------------------------------------------------
# 
#    Code to simulate post-inhibitory rebound coupled HH neurons.
# 
#    For Chapter 14, Section 14.7.2 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
# 
#    Written by James Keener and James Sneyd.
# 
#   -------------------------------------------------------------------

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint

# Global variables
gpir = 0.3
Vpir = 120
gL = 0.1
gsyn = 0.3
Vsyn = -80
phi = 3
thetasyn = -44
ksyn = 2
VL = -60

# Define the right-hand side of the ODE system
def rhs(s, t):
    V1, h1, V2, h2 = s

    minf1, hinf1, tauh1, Sinf1 = gates(V1)
    minf2, hinf2, tauh2, Sinf2 = gates(V2)

    V1p = -gpir * minf1**3 * h1 * (V1 - Vpir) - gL * (V1 - VL) - gsyn * Sinf2 * (V1 - Vsyn)
    h1p = phi * (hinf1 - h1) / tauh1
    V2p = -gpir * minf2**3 * h2 * (V2 - Vpir) - gL * (V2 - VL) - gsyn * Sinf1 * (V2 - Vsyn)
    h2p = phi * (hinf2 - h2) / tauh2

    return [V1p, h1p, V2p, h2p]

# Hodgkin-Huxley gate equations
def gates(V):
    minf = 1 / (1 + np.exp(-(V + 65) / 7.8))
    hinf = 1 / (1 + np.exp((V + 81) / 11))
    tauh = hinf * np.exp((V + 162.3) / 17.8)
    Sinf = 1 / (1 + np.exp(-(V - thetasyn) / ksyn))

    return minf, hinf, tauh, Sinf

# Initial conditions
init = [-50, 1, -40, 0]

# Time vector
t = np.arange(-50, 300, 0.1)

# Solve the ODE system
sol = odeint(rhs, init, t)

# Extract variables
V1, h1, V2, h2 = sol[:, 0], sol[:, 1], sol[:, 2], sol[:, 3]

# Plot results
plt.figure(figsize=(10, 6))
plt.plot(t, V1, label='V1')
plt.plot(t, V2, '--', label='V2')
plt.xlabel('Time')
plt.ylabel('Membrane Potential (mV)')
plt.legend()
plt.ylim(-80, -25)
plt.xlim(0, 300)
plt.show()
